pkg/controller/operators/olm/overrides/inject/inject.go

```diff
@@ -230,3 +230,79 @@ func InjectNodeSelectorIntoDeployment(podSpec *corev1.PodSpec, nodeSelector map[
 
 	return nil
 }
+
+// OverrideDeploymentAffinity will override the corev1.Affinity defined in the Deployment
+// with the given corev1.Affinity. Any nil top-level sub-attributes (e.g. NodeAffinity, PodAffinity, and PodAntiAffinity)
+// will be ignored. Hint: to overwrite those top-level attributes, empty them out. I.e. use the empty/default object ({})
+// e.g. NodeAffinity{}. In yaml:
+// affinity:
+//   nodeAffinity: {}
+//   podAffinity: {}
+//   podAntiAffinity: {}
+// will completely remove the deployment podSpec.affinity and is equivalent to
+// affinity: {}
+func OverrideDeploymentAffinity(podSpec *corev1.PodSpec, affinity *corev1.Affinity) error {
+	if podSpec == nil {
+		return errors.New("no pod spec provided")
+	}
+
+	if affinity == nil {
+		return nil
+	}
+
+	// special case: if podSpec.Affinity is nil or empty/default
+	// completely override podSpec.Affinity with overrides
+	if podSpec.Affinity == nil || reflect.DeepEqual(&corev1.Affinity{}, podSpec.Affinity) {
+		podSpec.Affinity = new(corev1.Affinity)
+		*podSpec.Affinity = *affinity
+		return nil
+	}
+
+	// special case: if overriding affinity is empty/default,
+	// nil out the deployment affinity
+	if reflect.DeepEqual(&corev1.Affinity{}, affinity) {
+		podSpec.Affinity = nil
+		return nil
+	}
+
+	// override non nil overriding fields
+	newAffinityValue := reflect.ValueOf(affinity)
+	curAffinityValue := reflect.ValueOf(podSpec.Affinity)
```

Comment:
Do we need to use `ValueOf` if these types are not interfaces? Couldn't we just dereference them normally?

Reply:
The point of this code was to handle API changes dynamically. I've updated to remove most of the reflection stuff and make it more readable/maintainable. I've added a unit test to detect whether there are been API changes and fail CI with a helpful message of what to do.